contracts/Staking.sol

```diff
@@ -458,66 +446,110 @@ contract Staking is Autopetrified, ERCStaking, ERCStakingHistory, IStakingLockin
+        account.totalLocked = account.totalLocked - lock_.amount;
 
-        // remove from active locks, replacing it by the last one in the array
-        // we assume consistency here, i.e., that lock exists in active array
-        uint256 locksLength = account.activeLockIds.length;
-        if (locksLength == 1) {
-            delete account.activeLockIds;
-            return;
-        }
+        emit Unlocked(_accountAddress, _lockManager, lock_.amount);
 
-        for (uint256 i = 0; i < locksLength; i++) {
-            if (account.activeLockIds[i] == _lockId) {
-                account.activeLockIds[i] = account.activeLockIds[locksLength - 1];
-                delete account.activeLockIds[locksLength - 1];
-                account.activeLockIds.length--;
-                break;
-            }
-        }
+        delete account.locks[_lockManager];
+    }
 
-        emit Unlocked(_accountAddress, _lockId, lock_.amount, lock_.manager, lock_.data);
+    function _increaseLockAllowance(address _lockManager, Lock storage _lock, uint256 _allowance) internal {
+        require(_allowance > 0, ERROR_AMOUNT_ZERO);
+
+        _lock.allowance = _lock.allowance.add(_allowance);
+
+        emit LockAllowanceChanged(msg.sender, _lockManager, _allowance, true);
     }
 
-    function _updateActiveLockAmount(address _accountAddress, uint256 _lockId, uint256 _amount, bool _increase) internal {
-        Lock storage lock_ = accounts[_accountAddress].locks[_lockId];
-        // check that lock hasn't been unlocked
-        require(lock_.unlockedAt > getTimestamp64(), ERROR_UNLOCKED_LOCK); // locks are created with a MAX_UINT64 unlockedAt
-        // checking that lock is in active array shouldn't be needed if data is consistent
+    function _increaseActiveLockAmount(address _accountAddress, address _lockManager, uint256 _amount) internal {
+        Account storage account = accounts[_accountAddress];
+        Lock storage lock_ = account.locks[_lockManager];
 
-        if (_increase) {
-            lock_.amount = lock_.amount.add(_amount);
+        uint256 newAmount = lock_.amount.add(_amount);
+        // check allowance is enough, it also means that lock hasn't been unlocked
+        require(newAmount <= lock_.allowance, ERROR_NOT_ENOUGH_ALLOWANCE);
+
+        lock_.amount = newAmount;
+
+        // update total
+        account.totalLocked = account.totalLocked.add(_amount);
+    }
+
+    function _decreaseActiveLockAmount(address _accountAddress, address _lockManager, uint256 _amount) internal {
+        Account storage account = accounts[_accountAddress];
+        Lock storage lock_ = account.locks[_lockManager];
+        // check that lock is enough, it also means that lock_.amount > 0 and therefore hasn't been unlocked
+        require(lock_.amount >= _amount, ERROR_NOT_ENOUGH_LOCK);
+
+        // if lock gets down to zero, unlock
+        if (lock_.amount == _amount) {
```

Comment:
Not sure I follow why we need this distinction here

Reply:
Yes, I think that doesn’t make sense anymore. Removing it.